core/services/service.go

```diff
@@ -75,11 +84,97 @@ import "context"
 type ServiceCtx interface {
 	// Start the service. Must quit immediately if the context is cancelled.
 	// The given context applies to Start function only and must not be retained.
+	//
+	// See MultiStart
 	Start(context.Context) error
 	// Close stops the Service.
 	// Invariants: Usually after this call the Service cannot be started
 	// again, you need to build a new Service to do so.
+	//
+	// See MultiCloser
 	Close() error
 
 	Checkable
 }
+
+// Group tracks a group of goroutines and provides shutdown signals.
+type Group struct {
+	wg sync.WaitGroup
+	utils.StopChan
+	logger.Logger
+}
+
+// Go calls fn in a tracked goroutine that will block closing the service.
+// fn should yield to StopChan promptly.
+func (g *Group) Go(fn func()) {
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		fn()
+	}()
+}
+
+// Spec specifies a service for New().
+type Spec struct {
+	Name        string
+	Start       func(context.Context) error
+	SubServices []ServiceCtx
+}
+
+type service struct {
+	utils.StartStopOnce
+	g    Group
+	spec Spec
+}
+
+// New returns a new ServiceCtx defined by Spec and a Group for managing goroutines and logging.
+// You *should* embed the ServiceCtx (to inherit methods), but *not* the Group:
+//
+//	type example struct {
+//		ServiceCtx
+//		g *Group
+//	}
+func New(spec Spec, lggr logger.Logger) (ServiceCtx, *Group) {
+	s := &service{
+		g: Group{
+			StopChan: make(utils.StopChan),
+			Logger:   lggr.Named(spec.Name),
+		},
+		spec: spec,
+	}
+	return s, &s.g
+}
+
+func (s *service) HealthReport() map[string]error {
+	m := map[string]error{s.Name(): s.Healthy()}
+	for _, sub := range s.spec.SubServices {
+		maps.Copy(m, sub.HealthReport())
+	}
```

Comment:
Instead of a doing a blind copy, which overwrites duplicate keys, we could check for dupes, and then report that here as an error :thinking: (or maybe just log it?)
I think that would catch most dupes :thinking: Things at the Register level are more obvious, but we could add the same kind of check there too.

Reply:
#10843